Fix setting boot path in case of UEFI partition (ESP) on MD RAID

pcahyna opened the pull request:
##### Pull Request Details:

* Type: **Bug Fix**

* Impact: **High**

* Reference to related issue (URL): #2595

* How was this pull request tested?
Backup & restore on a RHEL 8 host with UEFI but without RAID and on a CentOS 7 host with UEFI and RAID (ESP on RAID). Examination of logs to check that correct efibootmgr commands were executed.

* Brief description of the changes in this pull request:

Use dependencies when determining partitions for UEFI

The ESP may be located on a RAID device. In this case, we need to determine the physical RAID components and call efibootmgr on them (after restoring - during the finalize step), because efibootmgr does not know how to tell the firmware to boot from RAID. Moreover, the current code tries to parse a device like `/dev/md125` into a disk called `/dev/md` and partition `125`.

Fix by using storage dependencies: RAID components (the actual ESP copies) can be found as dependencies of the ESP block device. For consistency, use dependencies also to locate the ESP block device itself (the old method is preserved as a fallback).  Similar strategy is used by the GRUB2 installation code (finalize/Linux-i386/660_install_grub2.sh), which is able to install GRUB2 on the RAID components correctly.

In addition, clean up usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh a bit and add more logging to ease analysis when something goes wrong in this area.

TODO: finalize/Linux-i386/610_EFISTUB_run_efibootmgr.sh needs a similar fix.

Fixes #2595

jsmeix commented on usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh:
```diff
@@ -19,60 +19,60 @@ is_true $EFI_STUB && return
+# if mountpoint not found otherwise):
+if [ "$esp_mountpoint" = "/" ] ; then
+    esp_mountpoint="$TARGET_FS_ROOT/boot/efi"
+    Log "Mountpoint of $TARGET_FS_ROOT/$UEFI_BOOTLOADER not found, trying $esp_mountpoint"
```
Could you use `DebugPrint` (which also logs) instead of only `Log`
so that the message also appears on the user's terminal
when he runs recover in debug mode.
At least I prefer such messages where ReaR has to make special decisions
to also appear on the terminal because there it is easier to notice them
(instead of having to actively search for them from the log file).

pcahyna replied:
@jsmeix but `DebugPrint` also logs only when DEBUG is set, no? So if the user has not used debug mode, the message will go neither to the log not to the terminal. This does not seem right, because if we got here, we may encounter a potential problem later and the message in the log will be useful for debugging what has happened.

pcahyna replied:
We would need something like `LogDebugPrint`, but this function does not exist. I could use `LogPrint`, which would cause the message to go to log always and to terminal if verbose is set (instead of if debug is set). Would that be acceptable?

jsmeix replied:
In general I perfer too much verbosity over too less verbosity in verbose mode
(the former can be simply ignored, the latter leaves the user in the dark)
so yes, please use `LogPrint`.

Note that I wrote "in verbose mode" so by default ReaR should work silent
except real error messages but if ReaR's verbose mode is a bit too verbose
it is still better that too less verbose.

pcahyna replied:
I changed all `Log` to `LogPrint` in the fallback branches.

jsmeix commented on usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh:
```diff
@@ -19,60 +19,60 @@ is_true $EFI_STUB && return
+for efipart in $boot_efi_parts ; do
+    # /dev/sda1 or /dev/mapper/vol34_part2 or /dev/mapper/mpath99p4
+    Dev=$( get_device_name $efipart )
+    # 1 (must anyway be a low nr <9)
```
@pcahyna 
do you understand what that old comment
```
# 1 (must anyway be a low nr <9)
```
actually means?
I understand nothing at all what it means.

pcahyna replied:
@jsmeix 
> do you understand what that old comment
> 
> ```
> # 1 (must anyway be a low nr <9)
> ```
> 
> actually means?

No. it caught my attention as well, but I was afraid to remove it.

jsmeix replied:
LOL! :-))
I think I also have had a look at that comment some time ago.
I will dig a bit to find out if there is a meaning...
If not let's simply remove what nobody understands.

jsmeix replied:
I found where I had a look at that comment some time ago
https://github.com/rear/rear/commit/e3ca4ce8987dcd475696f32ce49ea3b0bc3dbb2a
and where that comment was kept but the matching code was changed
https://github.com/rear/rear/commit/bf6742bec5181ad7987a4a705668cc1a3834bef4
```
- ParNr=$( echo ${Dev} | sed -e 's/.*\([0-9]\).*/\1/' )  # 1 (must anyway be a low nr <9)
...
+ ParNr=$( get_partition_number $Dev )  # 1 (must anyway be a low nr <9)
```
and finally where that comment originated and where it made sense
https://github.com/rear/rear/commit/70490d16a62ee92a7129c5300a23eb29c22c2ef7
```
+ ParNr=$( echo ${Dev} | sed -e 's/.*\([0-9]\).*/\1/' )  # 1 (must anyway be a low nr <9)
```

So that comment has become obsolete since
https://github.com/rear/rear/commit/bf6742bec5181ad7987a4a705668cc1a3834bef4
and that comment should now be removed to avoid further confusion what it means.

pcahyna replied:
Thanks for the investigation. I removed the comment now.

jsmeix commented on usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh:
```diff
@@ -19,60 +19,60 @@ is_true $EFI_STUB && return
 test -f "$TARGET_FS_ROOT/$UEFI_BOOTLOADER" || return 0
 
```
Please add a
```
LogPrint "Creating EFI Boot Manager entries..."
```
here so that possibly subsequent messages appear in a meaningful context for the user
cf. finalize/Linux-i386/660_install_grub2.sh that also has such an initial `LogPrint`.

pcahyna replied:
Added.

jsmeix commented on usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh:
```diff
@@ -19,60 +19,60 @@ is_true $EFI_STUB && return
+    if ! test "$boot_efi_parts" ; then
+        LogPrintError "Cannot create EFI Boot Manager entry (unable to find partition for $boot_efi_dev)"
+        return 1
+    fi
```
Please add a `DebugPrint` message that informs the user
about the special case handling that ReaR had to do
when it was "unable to find ESP $esp_mountpoint_inside in layout"
e.g. something like
```
DebugPrint "Using fallback EFI boot partition(s) $boot_efi_parts (unable to find ESP $esp_mountpoint_inside in layout)"
```

pcahyna replied:
I used `Log` for now, will update together with the other `Log` calls when we agree on what should be the best replacement.

jsmeix replied:
Use `LogPrint` - see my other comment
https://github.com/rear/rear/pull/2608#discussion_r623785096

jsmeix commented on usr/share/rear/lib/layout-functions.sh:
```diff
@@ -427,6 +444,53 @@ get_partition_number() {
+# Should we use the information in $LAYOUT_DEPS, like get_parent_component does,
+# instead of string munging?
+function get_device_from_partition() {
+    local partition_block_device
```
@pcahyna 
thank you for the separated function to get the underlying device from a partition device.
It cleans up the code and makes it possible to easily improve the current
implementation (i.e. string munging) by something better in the future if needed.

jsmeix commented on usr/share/rear/lib/layout-functions.sh:
```diff
@@ -427,6 +444,53 @@ get_partition_number() {
+        device=${device%p}
+    fi
+
+    echo $device
```
What if `$device` is somehow invalid or empty?
Could some checking be implemented?
Currently `get_device_from_partition` is called only once
so the caller code could be enhanced to do some error handling
if `get_device_from_partition` returns non-zero exit value.

pcahyna replied:
@jsmeix

`$device` can't be empty here, it has been created in this function, unless the caller has provided an empty `$1`. Should we protect against that?

Concerning validity, we could check if the device actually exists and is a block device.

jsmeix replied:
You may catch empty `$1` after `partition_block_device=$1` with something like
```
contains_visible_char "$partition_block_device" || BugError "Empty string passed to get_device_from_partition"
```
or probably even better
```
test -b "$partition_block_device" || BugError "get_device_from_partition called with '$partition_block_device' that is no block device"
```
which also catches when `$1` is empty or only blanks and it shows that in the error message like
```
get_device_from_partition called with '' that is no block device
```
or
```
get_device_from_partition called with ' ' that is no block device
```

I tested that `test -b` behaves well also for symlinks
(at least on my openSUSE Leap 15.2 system):
```
# ls -l /dev/disk/by-id/wwn-0x5000039462b83c55-part1 | grep sda1
lrwxrwxrwx 1 root root 10 Apr 30 07:36 /dev/disk/by-id/wwn-0x5000039462b83c55-part1 -> ../../sda1

# test -b /dev/disk/by-id/wwn-0x5000039462b83c55-part1 && echo y || echo n
y
```

jsmeix replied:
Additionally the current "string munging" implementation
may result a `$device` string that is not a block device
so something like
```
test -b "$device" && echo $device
```
(`test -b "$device"` returns 1 when it is no block device)
as last command in the function
makes the function return 1 in case of errors
so that its caller can do whatever is best in its environment.

pcahyna replied:
Concerning invalid block devices, how does disk mapping work when block devices in the original system are mapped to different names in the rescue system? Is the layout and mapping regenerated to refer to the mapped names, or does it contain the original names and one has to map them after resolving dependences?
If the latter, it will have impact on the desirability of validation in this function, as it will be legitimate to handle device names that do not correspond to actual existing devices.

jsmeix replied:
Weekend is near so only a quick reply off the top of my head:

When things fail with errors because of disk mappings
it is still better than to silently proceed with wrong data.
I prefer user reports where things fail with clear error messages
over user reports when things fail in weird inexplicable ways,
cf. "Try hard to care about possible errors" in
https://github.com/rear/rear/wiki/Coding-Style

In migration mode "all needed files" (whatever that exactly means in practice ;-)
should be adapted during "rear recover".

The starting point is layout/prepare/default/300_map_disks.sh and
the actual mapping happens via layout/prepare/default/320_apply_mappings.sh
which contains
```
for file_to_migrate in "$LAYOUT_FILE" "$original_disk_space_usage_file" "$rescue_config_file" ; do
...
    ... apply_layout_mappings "$file_to_migrate" ...
```
so as far as it looks there mappings are only applied to
disklayout.conf layout/config/df.txt /etc/rear/rescue.conf

I assume those three files are usually sufficient
but I don't know for sure, in particular not for special cases.

jsmeix replied:
Searching the ReaR scripts for `apply_layout_mappings`
(that is the function that does the actual works)
finds that `apply_layout_mappings` is also called during "rear recover" in
finalize/GNU/Linux/250_migrate_disk_devices_layout.sh
and
finalize/GNU/Linux/260_rename_diskbyid.sh
where a lot more files (in particular restored bootloader config files)
get possibly adapted.

I am not a sufficient UEFI booting expert to know if there are missing
efibootmgr and/or EFI bootloader related files that would need to be adapted.

As far as I understand the code in finalize/Linux-i386/670_run_efibootmgr.sh
it seems disk mapping cannot affect how efibootmgr is called there
because it seems the code in finalize/Linux-i386/670_run_efibootmgr.sh
works directly on the actual state in the currently running recovery system because
it seems all is based on where ESP is mounted in the currently running recovery system
so nothing possibly outdated of how things have been on the original system is used.

pcahyna replied:
> As far as I understand the code in finalize/Linux-i386/670_run_efibootmgr.sh
> it seems disk mapping cannot affect how efibootmgr is called there
> because it seems the code in finalize/Linux-i386/670_run_efibootmgr.sh
> works directly on the actual state in the currently running recovery system because
> it seems all is based on where ESP is mounted in the currently running recovery system

After this proposed change, the code will use disk mapping (and not just the actual state). But, thanks to `apply_layout_mappings` being called earlier, the disk mappings should be in sync with the current disk names, so there should be no problem.

I am surprised a bit though that `apply_layout_mappings` gets called multiple times.

jsmeix replied:
Regarding "something is done multiple times" and similar things in ReaR:
Welcome to surprising adventures with grown code over many years by various contributors ;-)

I think ReaR is a perfect bazaar as in
https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar

pcahyna replied:
I added both checks that you suggested (for get_device_from_partition input and result).

jsmeix reviewed (approved):
From plain looking at the changes they look good to me
so I can approve it "bona fide" - in particular because things
are sufficiently commented so I think I even understand it a bit ;-)

pcahyna commented:
@gdha hello, should it have the **enhancement** label, when it Fixes #2595, which has the **bug** label?

pcahyna commented:
@rmetrich hello, please review, as you reported the problem.

rmetrich commented:
Hi @pcahyna I'm all good with your code, thanks a lot for the effort.

jsmeix commented:
I agree that ReaR is expected to restore a software RAID also on UEFI systesm
so the issue is a missing expected functionality a.k.a. "bug".

On the other hand there are zillions of other "reasonably expected" functionalitries
that are currently missing in ReaR so such issues are also enhancements
in particular when "reasonably expected" code for that is just not there.

For me a plain "bug" is only when code that is there does not work as intended.

I.e. it is also not a plain "bug" when code that is there does work as intended
but the current intention is not what others "reasonably expect" so such code
is "buggy" and needs some enhancement to behave as "reasonably expected".

jsmeix commented:
@pcahyna 
thank you so much for your additional clenaup work like your recent
https://github.com/rear/rear/pull/2608/commits/8228246d16b1374be5cc25fb7a1244c7102b804a

I do very much appreciate it!
It helps so much to get ReaR's code step by step into a better state
where others can easily understand WHY and WHAT and HOW things are done in ReaR
so that others can easily adapt and enhance ReaR in the future.

pcahyna commented:
@jsmeix, glad you like it! You know,
https://despair.com/products/quality

I can also replace the non-standard case identifiers like `Dev` and `ParNr` by something that conforms to the style guide, if you think that it makes sense to do it in the same PR.

jsmeix commented:
@pcahyna 
yes, feel free to mercilessly replace old non-standard identifiers with better names
in particular if needed with more explanatory names
cf. "Code must be easy to read" in
https://github.com/rear/rear/wiki/Coding-Style

I appreciate it!

jsmeix commented:
Some "famous words" before weekend regarding
https://despair.com/products/quality
```
The race for quality has no finish line- so technically, it's more like a death march.
```

Yes, we all will be dead at some time.

But if we made code as intended by
https://github.com/rear/rear/wiki/Coding-Style
our code can evolve (it may not evolve but it can evolve)
so from our code future code can be derived (this may not happen but it can happen)
in contrast to code that (amost) nobody understands which will be completely removed
and replaced by other completely new code from scratch
so code that is hard to understand dies out
while code that is easy to understand can have descendants.

gozora reviewed (approved):
From the plain looking at the code, it looks good to me.

V.

jsmeix commented:
It seems there is a dilemma where in the code
the `Creating EFI Boot Manager entries...` info
should be done.

Current finalize/Linux-i386/670_run_efibootmgr.sh code in this pull request is (excerpts)
```
LogPrint "Creating EFI Boot Manager entries..."
...
if [ "$esp_mountpoint" = "/" ] ; then
    esp_mountpoint="$TARGET_FS_ROOT/boot/efi"
    Log "Mountpoint of $TARGET_FS_ROOT/$UEFI_BOOTLOADER not found, trying $esp_mountpoint"
fi
...
test -d "$esp_mountpoint" || return 0
```
so it could tell the user `Creating EFI Boot Manager entries...`
but then it does nothing.

I think when `test -d "$esp_mountpoint"` fails is must no longer exit silently
so a matching message should be added like
```
if ! test -d "$esp_mountpoint" ; then
    LogPrint "Skip creating EFI Boot Manager entries (no ESP mountpoint found)"
fi
```

pcahyna commented:
> I think when `test -d "$esp_mountpoint"` fails is must no longer exit silently

This part caught my attention as well. Is it even wise to return 0 in this case? At this point we know that a UEFI bootloader should be used, so skipping the rest of the file is almost certainly a serious error.
(Not so much related to the subject of the PR anymore though. For now, I can add the other LogPrint as you suggest.)

pcahyna commented:
> For now, I can add the other LogPrint as you suggest

Done (I changed the message a bit.)

jsmeix commented on usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh:
```diff
@@ -32,7 +32,10 @@ fi
 # Double quotes are mandatory here because 'test -d' without any (possibly empty) argument results true:
-test -d "$esp_mountpoint" || return 0
+if ! test -d "$esp_mountpoint" ; then
+    LogPrint "Skip creating EFI Boot Manager entries (ESP mountpoint $esp_mountpoint does not exist)"
```
Nitpicking:
Strictly speaking $esp_mountpoint may exist but is no directory.

pcahyna replied:
I changed it to
"(no ESP mountpoint directory $esp_mountpoint)"

jsmeix commented:
@pcahyna 
regarding your https://github.com/rear/rear/pull/2608#issuecomment-831263524
```
At this point we know that a UEFI bootloader should be used,
so skipping the rest of the file is almost certainly a serious error.
```

Then show it as an error to the user so he is informed something went wrong
(e.g. use wording like `failed to ...` or `cannot ...` that tells something went wrong)
and the user knows he may have to fix something manually after "rear recover"
and leave the script with `return 1` (the `Source` function doesn't abort)
but do not error out of "rear recover", cf. finalize/Linux-i386/660_install_grub2.sh
https://github.com/rear/rear/blob/master/usr/share/rear/finalize/Linux-i386/660_install_grub2.sh

pcahyna commented:
> Then show it as an error to the user so he is informed something went wrong
> (e.g. use wording like `failed to ...` or `cannot ...` that tells something went wrong)
> and the user knows he may have to fix something manually after "rear recover"
> and leave the script with `return 1`

I can do that, but I feel this deviates a lot from the subject of this PR (the questionable return 0 is inherited from the original code).

pcahyna commented:
@jsmeix I believe that I addressed all your comments except for the `return 0` in some places where it had existed before my changes. Please review again. If everything is ok, I will force push to squash some commits together.

pcahyna commented:
> @pcahyna
> yes, feel free to mercilessly replace old non-standard identifiers with better names
> in particular if needed with more explanatory names
> cf. "Code must be easy to read" in
> https://github.com/rear/rear/wiki/Coding-Style
> 
> I appreciate it!

One of the last commits does that now.

jsmeix commented:
@pcahyna 
thank you for all your work here - in particular for your additional cleanup work!

jsmeix commented:
@rear/contributors 
if there are no objections I would like to merge it tomorrow afternoon.

jsmeix commented:
@pcahyna 
the `return 0` when there is `no ESP mountpoint directory $esp_mountpoint` was from me via
https://github.com/rear/rear/commit/cb6fe3be719c1361b22d0484cd940c681752a973

I am sure I could not think about if `return 1` could be better for this particular case
because at that time I did not know your reasoning why `return 1` is right here in
https://github.com/rear/rear/pull/2608#issuecomment-831263524

pcahyna commented:
@jsmeix are there any guidelines for return codes from scripts? ReaR does not seem to care that much, actually. There is just `Debug "Source function: 'source $source_file' returns $source_return_code"` in the `Source()` function. So it looks like returning 1 when problems are discovered is not enough to get them noticed by the user and one should complain very loudly (`LogPrintError`, probably).

pcahyna commented on usr/share/rear/finalize/Linux-i386/670_run_efibootmgr.sh:
```diff
@@ -18,61 +20,72 @@ is_true $EFI_STUB && return
-        (*)      Log "Unsupported kpartx partition delimiter for $Dev"
-    esac
+if ! test -d "$esp_mountpoint" ; then
+    LogPrint "Skip creating EFI Boot Manager entries (no ESP mountpoint directory $esp_mountpoint)"
```
I should use `LogPrintError` instead of `LogPrint` here, and return 1 ?

pcahyna replied:
also, reword as "failed to" or "error ..." or something similar instead of just "Skip".

jsmeix replied:
Yes, I would much appreciate if you could use `LogPrintError` and `return 1`
plus more explicit wording in the message that something had failed
in this particular case.

pcahyna replied:
Done. Also for the other `return` above

jsmeix commented:
@pcahyna 
yes, what you describe in your
https://github.com/rear/rear/pull/2608#issuecomment-831916709
is exactly the current state in ReaR's code.

pcahyna commented:
@jsmeix please check if you approve the latest change, esp. the wording of the error messages. If yes, I consider this PR functionally complete and I will rebase and force-push the branch to squash related commits together.

jsmeix commented:
@pcahyna 
thank you for your final adaptions and enhancements.
Even if those kind of smaller things might look unimportant or nitpicking
I appreciate them so much because at least for me such smaller things are important
because they can make a bigger difference how others can understand what goes on.

I dated to do some further tiny adaptions and enhancements regarding such smaller things
via my recent commits here - I hope that is OK for you @pcahyna 

Now all looks very well to me
and I would like to merge it today afternoon
unless there are objections from other @rear/contributors

jsmeix commented:
@pcahyna 
no need to rebase and force-push the branch to squash related commits together.
You can just keep the individual development commits as is.
The merge commit will automatically be the one that has all of them in one commit.
I even prefer when all the individual development commits are kept as is.
It preserves the development history in the master branch
i.e. I perfer to not remove information unless there is a serious reason to do so.

In my local `git clone https://github.com/rear/rear.git` I often use
```
git log --format="%ae %H %ad%n%s :%n%b%n" --graph | fmt -w 120 -u -t | less
```
to show the whole git history in the master branch and there
all the individual development commits often help me
to better understand in retrospect why things were changed
without the need to dig out the GitHub issue and pull request
and read through all the details there.

Having all the individual development commits in the master branch history
is often even crucial to understand in retrospect why things were changed
because the merge commit message is often only rather meaningless, e.g.:
```
* | | | gratien.dhaese@gmail.com 285c0c7660caca184ea53dbec5525a69af1fdcc9 Tue Jan 26 14:44:39 2021 +0100
|\ \ \ \ Merge pull request #2544 from pcahyna/netbackup-vxpbx_exchanged :
| | | | | Changes for NetBackup (NBU) support
| | | | |
| * | | | pcahyna@redhat.com 9a0fcbfd0dcfecdfce5921e24a89e3918d962467 Wed Dec 16 11:08:14 2020 +0100
| | | | | Changes for NetBackup (NBU) support :
| | | | | Add /usr/openv/tmp directory to the NBU skeleton
| | | | |
| | | | |
| * | | | pcahyna@redhat.com 0c6909ed2c37dfa92f87b0714b62af7cae4720c9 Wed Dec 16 10:40:55 2020 +0100
| |/ / / Changes for NetBackup (NBU) support :
| | | | Copy NetBackup PBX related files to the rescue system and start
| | | | vxpbx_exchanged on boot
```
The merge commit message `Changes for NetBackup (NBU) support` alone
tells nothing what was changed.

pcahyna commented:
@jsmeix 

your newest commits are ok. But I don't entirely agree with
> no need to rebase and force-push the branch to squash related commits together.
> Just keep the individual development commits as is.

There are commits marked as squash! and fixup! and those were intended to be squashed with others. My intent was to preserve the history, but not all of the commits, there were commits that merely amend earlier commits and the history would IMO look better with them squashed into earlier commits. The squash! and fixup! log messages will look particularly ugly in the history because they were not intended to be part of the final history.

jsmeix commented:
@pcahyna 
feel free to squash your commits as you like
provided the commit messages of the commits you finally keep
still show sufficient information so that others can understand in retrospect
what was changed without the need to dig out the GitHub issue
and pull request and read through all the details there.

In particular we need that understanding in retrospect
when we make the rear/doc/rear-release-notes.txt
cf. "compile the information for the release notes" in
https://github.com/rear/rear/wiki/Release-process

pcahyna commented:
@jsmeix rebase done.

jsmeix commented:
@pcahyna 
thank you.

I will merge it soon today.

pcahyna commented:
Oops. I accidentally made the file executable in 27de4cd95dabe86f531f132108744459b2dbcfd0 @jsmeix can you please revert the mode change, or should I submit a new PR reverting it?

jsmeix commented:
@pcahyna regarding executable finalize/Linux-i386/670_run_efibootmgr.sh
you could simply submit it with mode `-rw-r--r--` as "by the way" addon
via your next PR without mentioning it in a commit message
(but you can mention it in a separated PR comment).
